server/src/test/java/org/candlepin/resteasy/JsonProviderTest.java

```diff
@@ -68,6 +81,45 @@ public void serializedDateDoesNotIncludeMilliseconds() throws JsonProcessingExce
         assertTrue(serializedDate.equals(expectedDate));
     }
 
+    /*
+     * This tests to see that our DateSerializer does not fail with thread-safety issues caused by
+     * SimpleDateFormat. As the https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html doc
+     * explains: "Date formats are not synchronized. It is recommended to create separate format instances
+     * for each thread. If multiple threads access a format concurrently, it must be synchronized externally."
+     *
+     * For this reason, SimpleDateFormat has been replaced with DateTimeFormatter, which is thread-safe.
+     */
+    @Test
+    public void canSerializeDatesWithLargeTimestampsConcurrently() {
+        JsonProvider provider = new JsonProvider(config,
+            new ProductCachedSerializationModule(productCurator));
+        ObjectMapper mapper = provider.locateMapper(Object.class, MediaType.APPLICATION_JSON_TYPE);
+
+        ExecutorService ex = Executors.newFixedThreadPool(1000);
+        for (int i = 0; i < 500000; i++) {
+            ex.execute(() -> {
+                try {
+                    mapper.writeValueAsString(new Date(getRandomTimestampWithMilliSeconds()));
```

Comment:
Should we also verify that the value written is as expected?

Reply:
Done. This is actually good because running this test with the old code, there's quite a few dates that are not serialized accurately (I suspect because of similar threading issues), while the new code fixes those (all dates are now serialized accurately).